synapse/api/errors.py

```diff
@@ -307,6 +313,32 @@ def __init__(
         super().__init__(code, msg, errcode, additional_fields)
 
 
+class UnstableSpecAuthError(AuthError):
+    """An error raised when a new error code is being proposed to replace a previous one.
+    This error will return a "org.matrix.unstable.errcode" property with the new error code,
+    with the previous error code still being defined in the "errcode" property
+    """
+
+    def __init__(
+        self,
+        code: int,
+        msg: str,
+        errcode: str,
+        previous_errcode: str = Codes.FORBIDDEN,
+        additional_fields: Optional[dict] = None,
+    ):
+        self.previous_errcode = previous_errcode
+        super().__init__(code, msg, errcode, additional_fields)
+
+    def error_dict(self) -> "JsonDict":
+        return cs_error(
+            self.msg,
+            self.previous_errcode,
+            **{"org.matrix.unstable.errcode": self.errcode},
```

Comment:
Should this key also not be namespaced to the MSC?

Reply:
It could be, my main concern was the maintenance burden it puts on the client (and servers) to maintain a set of keys **and** values to namespace, which makes things like these error classes more cumbersome (you could argue there is a *benefit* to making unstable things more cumbersome though).

If we think there is a real danger of clashing with other MSCs or it just being too unusual, I can namespace them.

Reply:
Well, future MSCs could reuse the MSC namespaced key, it just makes it explicitly clear where to go to find out what the key *means*. (Otherwise different MSCs might use subtly different definitions of the key).

Reply:
Done in d1a66cb